### ISSUE TITLE
Non-portable shebang line

### ISSUE THREAD

**Issue by vendion** [TheLocehiliosan/yadm#14]
The shebang line in yadm won't work for all systems as some unixes and unix like OSes put `bash` in `/bin/bash`.  For example openBSD puts `bash` in `/usr/local/bin/bash` instead of `/bin/bash` as `bash` is something the user has to explicitly install, not provided by default install.  The recommended work around for this is to use `#!/usr/bin/env bash` instead as `env` is placed as `/usr/bin/env` in more operating systems than `bash` is put in `/bin`.

Details: http://stackoverflow.com/a/10383546/2524942 and https://en.wikipedia.org/wiki/Shebang_%28Unix%29#Portability

**Comment by DoomHammer** [comment]
Yes, but only to a certain degree. There are systems where `env` is not located in `/usr/bin` but in `/bin`. Sometimes one is a symlink to the other, sometimes only one of them exists.

So far I haven't found a better way to have a portable shebang other than using `#!/bin/sh`.

**Comment by TheLocehiliosan** [comment]
I am interested in making **yadm** as portable as it can be. Of course using `/bin/sh` might be the most portable, but I don't want to give up Bash conveniences.

There are a limited number of systems which I can test with. Below are systems I have access to, and the location of `env` and `bash` for them. Using `#!/usr/bin/env bash` does appear to work with the systems below. I'm also assuming (based on the comment from @vendion) that it works on OpenBSD as well—if Bash is installed.

| Distro    | `env` path                   | `bash` path |
| --------- | ---------------------------- | ----------- |
| CentOS 5  | `/usr/bin/env` -> `/bin/env` | `/bin/bash` |
| CentOS 6  | `/usr/bin/env` -> `/bin/env` | `/bin/bash` |
| CentOS 7  | `/usr/bin/env`               | `/bin/bash` |
| Fedora 23 | `/usr/bin/env`               | `/bin/bash` |
| OS X      | `/usr/bin/env`               | `/bin/bash` |
| Debian 8  | `/usr/bin/env`               | `/bin/bash` |
| Ubuntu 15 | `/usr/bin/env`               | `/bin/bash` |

I don't know the full set of systems that users are using **yadm** with. This is the first time this issue has been raised. It would be helpful if users could comment on other systems they use **yadm** with, what the path for `env` is and if a proper version of Bash is located in the system's path.

@DoomHammer, can you list the systems you know that do not have `env` located in `/usr/bin`?

**Comment by vendion** [comment]
@TheLocehiliosan That is a nice chart, here is the systems that I know as well:

OS | `env` path | `bash` path
---- | -------------- | ----------------
openSUSE 13.1 | `/usr/bin/env` | `/bin/bash`
FreeBSD 10.3 | `/usr/bin/env` | `/usr/local/bin/bash`
OpenBSD 5.9 | `/usr/bin/env` | `/usr/local/bin/bash`
PacBSD | `/usr/bin/env` | `/usr/bin/bash`
Arch Linux | `/usr/bin/env` | `/usr/bin/bash`
Minix 3.3* | `/usr/bin/env` | `/usr/pkg/bin/bash`

\* Because Minix 3.3 is heavily based on NetBSD, I assume it will be the same for them but don't have access to a NetBSD box to test.

@DoomHammer True that not every Unix-based system puts `env` in `/usr/bin` nor does every one of them symlink it there (I think GNU Hurd is one such system but haven't tested it), but the number of systems that do is much higher than the ones that put `bash` in `/bin`.  So while using a shebang line like `#!/usr/bin/env bash` wouldn't be as perfectly portable it is more portable than `#!/bin/bash` without needing to go to bourne shell (`/bin/sh`)

*Disclaimer*: I don't yet use **yadm**, due to this problem, but looks very interesting and found out about this thanks to a post in [/r/unixporn](https://www.reddit.com/r/unixporn).  I'm currently using **vcsh** which does use `/bin/sh` for its shebang line, so avoids portability issues that way.

**Comment by DoomHammer** [comment]
`busybox` based systems tend to not have `/usr/bin` at all. I understand this is a rare issue, but worth thinking about as well.

As for the `#!/bin/sh` shebang I suggested it does not necessarily mean giving up `bash` altogether! You can simple have a portable wrapper that would:

```
#!/bin/sh
env bash /path/to/script.bash
```

or

```
#!/bin/sh
bash <<'END'
'END'
```

**Comment by vendion** [comment]
As long as the arguments sent to the wrapper are passed to the bash script I don't see a reason why that wouldn't work.

**Comment by TheLocehiliosan** [comment]
I believe `-s` is required to properly obtain the arguments passed to **yadm**.

```
#!/bin/sh
bash -s "$@" <<'END'
<code>
'END'
```

There are at least two frustrating downside to this for development/maintenance:
* Syntax highlighting will see the entire program as a static string
* Automated [shellcheck](http://www.shellcheck.net) testing doesn't test the program, it only checks the very simple wrapper

I suppose I could also have build process which merely "wraps" the program when releasing it.

The other suggestion, `env bash /path/to/script.bash`, isn't without downsides either. Doing that requires _two_ files be installed. And it requires a hardcoded path to where the actual program file.

I'm curious if there are any other suggestions/comments about for making this the most portable, without much pain.

**Comment by DoomHammer** [comment]
I agree shipping two files is an overkill.

What about [this approach](http://stackoverflow.com/a/19538774)?

**Comment by TheLocehiliosan** [comment]
Thanks. Checking `$BASH_VERSION` and conditionally `exec`ing **bash** seems like it could be a good approach. I've pushed a version like that to the dev branch. I'm curious if anyone else has other input.

**Comment by irl** [comment]
From dealing with Debian packaging, I find more and more /usr/bin/env being the best solution. This even works if the system didn't have bash on it but the user has installed bash in ~/bin/bash.

I would also add that for distribution packaging, this is something that is trivially fixed by a patch by the distribution developers if /usr/bin/env doesn't exist.

**Comment by DoomHammer** [comment]
@irl I think you are right if we are talking about packaged apps. Personally, I see `yadm` as a ready-to-use script rather than a packaged software. Hence the ability to run it as-is should probably be more important.

Linuxbrew is a similar type of software (curl && run) and it uses the `exec bash` method. The reasoning is thus:

1. `/usr/bin/env` may be `/bin/env`
2. `/bin/bash` may be `/usr/bin/bash` or `/usr/local/bin/bash` or [...]
3. `/bin/sh` is almost always true.
4. with proper `$PATH` an appropriate `bash` will get `exec`d unless the script is already running under `bash`.

**Comment by TheLocehiliosan** [comment]
Checking `$BASH_VERSION` and conditionally execing `bash` seems very solid, and doesn't require packagers to do anything. It also works if a system doesn't have bash, but a user has it installed in their path. I plan to include this change in version **1.05**.

**Comment by vendion** [comment]
Some distributions prefer to not patching software unless absolutely needed that way they stay as close to the way upstream intended it to work (See https://wiki.archlinux.org/index.php/Arch_Linux#Simplicity).  Plus something like this I see no reason why it can't be address upstream rather than having the people that maintain the different Linux/BSD/Solaris/etc packages, by addressing upstream it allow the project to work in a "ready-to-use" manner where the user can just download and run the script.  

Checking `$BASH_VERSION` then conditionally execing `bash` has been working without issue in all the use cases I have tested, so I'm good with that approach.
